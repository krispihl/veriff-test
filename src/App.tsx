import React, { useEffect, useState, useCallback } from "react";
import { fetchChecks, submitCheckResults } from "./api";
import { ListItem } from "./components/ListItem";
import { Button } from "./components/Button";
import { ErrorView } from "./views/Error";
import { SuccessView } from "./views/Success";
import { Check } from './types'
import "./global.css";

const App = () => {
	const [checks, setChecks] = useState<Check[]>([]);
	const [isLoading, setIsLoading] = useState(true);
	const [isSuccess, setIsSuccess] = useState(false);
	const [isError, setIsError] = useState(false);
	const [buttonDisabled, setButtonDisabled] = useState(true);

	// Fetch data with API call
	useEffect(() => {
		const fetchData = async () => {
			try {
				const response: Check[] = await fetchChecks();

				const sorted = response.sort((a, b) => a.priority - b.priority);
				const modified = sorted.map((check, idx) => {
					return {
						...check,
						active: false,
						disabled: idx === 0 ? false : true,
					};
				});

				setChecks(modified);
			} catch (error) {
				setIsError(true);
			}

			setIsLoading(false);
		};

		fetchData();
	}, []);

	// Update checks based on user keypress events
	const handleArrowUp = useCallback(
		(activeItemIndex: number) => {
			if (activeItemIndex !== 0) {
				const modified = checks.map((check, idx) => {
					return {
						...check,
						active: idx === activeItemIndex - 1 ? true : false,
					};
				});

				setChecks(modified);
			}
		},
		[checks]
	);

	const handleArrowDown = useCallback(
		(activeItemIndex: number) => {
			if (activeItemIndex !== checks.length - 1 && !checks[activeItemIndex + 1].disabled) {
				const modified = checks.map((check, idx) => {
					return {
						...check,
						active: idx === activeItemIndex + 1 ? true : false,
					};
				});

				setChecks(modified);
			}
		},
		[checks]
	);

	const handleOne = useCallback(
		(activeItemIndex: number) => {
			const modified = checks.reduce((previousValue, currentValue, currentIndex) => {
				const modifiedCheck = [{
					...currentValue,
					value: currentIndex === activeItemIndex ? 'yes' : currentValue.value,
					disabled:
					currentIndex <= activeItemIndex ||
					currentIndex === activeItemIndex + 1 ||
					// handling edge case when user goes back and changes one of the previous values to NO and then back to YES
					(currentIndex >= activeItemIndex + 2 && previousValue[currentIndex - 1].value === 'yes' && !previousValue[currentIndex - 1].disabled)
						? false
						: true,
				}]

				return previousValue.concat(modifiedCheck);
			}, [] as Check[]);

			setChecks(modified);
		},
		[checks]
	);

	const handleTwo = useCallback(
		(activeItemIndex: number) => {
			const modified = checks.map((check, idx) => {
				return {
					...check,
					value: idx === activeItemIndex ? 'no' : check.value,
					disabled: idx <= activeItemIndex ? false : true,
				};
			});

			setChecks(modified);
		},
		[checks]
	);

	const handleKeyPress = useCallback(
		(event) => {
			if(['ArrowUp', 'ArrowDown'].indexOf(event.code) > -1) {
				event.preventDefault();
			};

			const activeItemIndex = checks.findIndex((check) => check.active);

			switch (event.key) {
				case 'ArrowUp':
					handleArrowUp(activeItemIndex);
					break;
				case 'ArrowDown':
					handleArrowDown(activeItemIndex);
					break;
				case '1':
					handleOne(activeItemIndex);
					break;
				case '2':
					handleTwo(activeItemIndex);
					break;
				default:
					return;
			}
		},
		[checks, handleArrowUp, handleArrowDown, handleOne, handleTwo]
	);

	// Attach event listener for keydown event
	useEffect(() => {
		document.addEventListener('keydown', handleKeyPress);

		return () => {
			document.removeEventListener('keydown', handleKeyPress);
		};
	}, [handleKeyPress]);

	// Update checks based on user click event
	const updateChecks = useCallback(
		(value: string, id: string) => {
			const activeItemIndex = checks.findIndex((x) => x.id === id);

			const updatedChecks = checks.reduce((previousValue, currentValue, currentIndex) => {
				const modifiedCheck = [{
					...currentValue,
					active: currentIndex === activeItemIndex ? true : false,
					disabled:
					currentIndex <= activeItemIndex ||
					(currentIndex === activeItemIndex + 1 && value === 'yes') ||
					// handling edge case when user goes back and changes one of the previous values to NO and then back to YES
					(currentIndex >= activeItemIndex + 2 && previousValue[currentIndex - 1].value === 'yes' && !previousValue[currentIndex - 1].disabled)
						? false
						: true,
					value: currentIndex === activeItemIndex ? value : currentValue.value,
				}]

				return previousValue.concat(modifiedCheck);
			}, [] as Check[]);

			setChecks(updatedChecks);
		},
		[checks]
	);

	// Submit button logic
	useEffect(() => {
		setButtonDisabled(true);
		const values = checks.map((check) => check.value);

		const allChecksYes = values.length && values.every((value) => value && value === 'yes');
		const someCheckNo = values.includes('no');

		if (allChecksYes || someCheckNo) {
			setButtonDisabled(false);
		}
	}, [checks]);

	// Submit check results
	const handleSubmit = useCallback(async () => {
		const submitData = checks.map((check) => {
			return {
				checkId: check.id,
				value: check.value || 'not selected',
			};
		});

		try {
			await submitCheckResults(submitData);
			setIsSuccess(true);

			console.log('submitData', submitData);
		} catch (error) {
			setIsError(true);
		}
	}, [checks]);

	return (
		<div className='app'>
			{isLoading && <p className='loadingState'>Loading...</p>}
			{isSuccess && <SuccessView />}
			{isError && <ErrorView />}
			{!isLoading && !isSuccess && !isError && (
			<div className='wrapper'>
				<form>
					{checks.map((check, idx) => {
						return (
							<ListItem
								key={check.id}
								check={check}
								updateChecks={updateChecks}
							/>
						);
					})}
				</form>
				<Button disabled={buttonDisabled} onClick={handleSubmit}>
					Submit
				</Button>
			</div>
			)}
		</div>
	);
};

export default App;
