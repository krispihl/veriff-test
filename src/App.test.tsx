import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';
import { fetchChecks, submitCheckResults } from "./api";

const mockedFetchChecks = fetchChecks as jest.Mocked<any>;
const mockedSubmitCheckResults = submitCheckResults as jest.Mocked<any>;

jest.mock('./api', () => ({
    fetchChecks: jest.fn(),
    submitCheckResults: jest.fn(),
}))

describe('App', () => {
    beforeEach(() => {
        mockedFetchChecks.mockResolvedValue([
            {
                id: 'aaa',
                priority: 10,
                description: 'Face on the picture matches face on the document'
            },
            {
                id: 'bbb',
                priority: 5,
                description: 'Veriff supports presented document'
            },
        ]);
    })

    it('renders App correctly when checks are fetch successfully', async () => {
        const { container } = render(<App />);

        expect(await screen.findByText('Veriff supports presented document')).toBeInTheDocument();
        expect(await screen.findByText('Face on the picture matches face on the document')).toBeInTheDocument();
        expect(await screen.findByText('Submit', { exact: true})).toBeDisabled();
        
        expect(container).toMatchSnapshot();
    });

    it('renders Error view when unable to fetch checks', async () => {
        mockedFetchChecks.mockRejectedValue({ success: false });
        const { container } = render(<App />);

        expect(await screen.findByText('Whoops! Something went wrong...')).toBeInTheDocument();        
        expect(container).toMatchSnapshot();
    });

    it('renders App correctly when user makes a selection', async () => {
        const { container } = render(<App />);

        const buttons = await screen.findAllByRole('button');

        // user clicks first ListItem button YES
        userEvent.click(buttons[0]);

        expect(container).toMatchSnapshot();
        expect(await screen.findByText('Submit', { exact: true})).toHaveAttribute('disabled');
        expect(buttons[0]).toHaveClass('activeButton');
        expect(buttons[1]).not.toHaveClass('activeButton');

        // user clicks first ListItem button NO
        userEvent.click(buttons[1]);

        expect(buttons[1]).toHaveClass('activeButton');
        expect(buttons[0]).not.toHaveClass('activeButton');
        expect(await screen.findByText('Submit', { exact: true})).not.toHaveAttribute('disabled');
    });

    it('sends submit data when user clicks Submit button and renders success view', async () => {
        const { container } = render(<App />);

        const buttons = await screen.findAllByRole('button');

        // user clicks first ListItem button NO
        userEvent.click(buttons[1]);
        // user clicks Submit button
        userEvent.click(buttons[4]);

        expect(await screen.findByText('Submit', { exact: true})).toBeEnabled();
        expect(mockedSubmitCheckResults).toHaveBeenCalledTimes(1);
        expect(mockedSubmitCheckResults).toHaveBeenCalledWith([
            {
                checkId: 'bbb',
                value: 'no',
            },
            {
                checkId: 'aaa',
                value: 'not selected',
            },
        ]);

        expect(container).toMatchSnapshot();
        expect(await screen.findByText('Great success!')).toBeInTheDocument();
    });

    it('renders App correctly when user navigates with keyboard', async () => {
        const { container } = render(<App />);

        const buttons = await screen.findAllByRole('button');
        const listItems = await screen.findAllByTestId('list-item')

        // user selects first listItem
        userEvent.keyboard('[ArrowDown]');

        // first listItem becomes active, second is still disabled
        expect(listItems[0]).toHaveClass('item--active');
        expect(listItems[1]).toHaveClass('item--disabled');

        // user uses right arrow key, nothing changes
        userEvent.keyboard('[ArrowRight]');

        expect(listItems[0]).toHaveClass('item--active');
        expect(listItems[1]).toHaveClass('item--disabled');

        // user selects first listItem button YES
        userEvent.keyboard('[Digit1]');

        expect(buttons[0]).toHaveClass('activeButton');
        //second listItem is not disabled anymore
        expect(listItems[1]).not.toHaveClass('item--disabled');
        // submit button is still disabled
        expect(buttons[4]).toBeDisabled();
        expect(container).toMatchSnapshot();

        // user selects second listItem
        userEvent.keyboard('[ArrowDown]');

        // second listItem becomes active, first listItem will not be active anymore
        expect(listItems[1]).toHaveClass('item--active');
        expect(listItems[0]).not.toHaveClass('item--active');

        // user selects again first listItem
        userEvent.keyboard('[ArrowUp]');

        // first listItem becomes active, second listItem will not be active anymore
        expect(listItems[0]).toHaveClass('item--active');
        expect(listItems[1]).not.toHaveClass('item--active');

        // user selects again second listItem
        userEvent.keyboard('[ArrowDown]');
        // user selects second listItem button NO
        userEvent.keyboard('[Digit2]');

        expect(buttons[3]).toHaveClass('activeButton');
        // submit button will now be enabled
        expect(buttons[4]).toBeEnabled();
    });

    it('renders Error view when sumbit fails', async () => {
        mockedSubmitCheckResults.mockRejectedValue({ success: false });

        const { container } = render(<App />);
        const buttons = await screen.findAllByRole('button');

        expect(await screen.findByText('Submit', { exact: true})).toBeDisabled();

        userEvent.click(buttons[1]);
        userEvent.click(buttons[4]);
        
        expect(await screen.findByText('Submit', { exact: true})).toBeEnabled();
        expect(container).toMatchSnapshot();
        expect(await screen.findByText('Whoops! Something went wrong...')).toBeInTheDocument();
    });
})