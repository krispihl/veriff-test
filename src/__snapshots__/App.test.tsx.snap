// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders App correctly when checks are fetch successfully 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="wrapper"
    >
      <form>
        <div
          class="item"
          data-testid="list-item"
        >
          <p
            class="text"
          >
            Veriff supports presented document
          </p>
          <div>
            <button
              class="leftButton"
            >
              Yes
            </button>
            <button
              class="rightButton"
            >
              No
            </button>
          </div>
        </div>
        <div
          class="item item--disabled"
          data-testid="list-item"
        >
          <p
            class="text"
          >
            Face on the picture matches face on the document
          </p>
          <div>
            <button
              class="leftButton disabled"
            >
              Yes
            </button>
            <button
              class="rightButton disabled"
            >
              No
            </button>
          </div>
        </div>
      </form>
      <button
        class="button"
        disabled=""
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;

exports[`App renders App correctly when user makes a selection 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="wrapper"
    >
      <form>
        <div
          class="item item--active"
          data-testid="list-item"
        >
          <p
            class="text"
          >
            Veriff supports presented document
          </p>
          <div>
            <button
              class="leftButton activeButton"
            >
              Yes
            </button>
            <button
              class="rightButton"
            >
              No
            </button>
          </div>
        </div>
        <div
          class="item"
          data-testid="list-item"
        >
          <p
            class="text"
          >
            Face on the picture matches face on the document
          </p>
          <div>
            <button
              class="leftButton"
            >
              Yes
            </button>
            <button
              class="rightButton"
            >
              No
            </button>
          </div>
        </div>
      </form>
      <button
        class="button"
        disabled=""
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;

exports[`App renders App correctly when user navigates with keyboard 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="wrapper"
    >
      <form>
        <div
          class="item item--active"
          data-testid="list-item"
        >
          <p
            class="text"
          >
            Veriff supports presented document
          </p>
          <div>
            <button
              class="leftButton activeButton"
            >
              Yes
            </button>
            <button
              class="rightButton"
            >
              No
            </button>
          </div>
        </div>
        <div
          class="item"
          data-testid="list-item"
        >
          <p
            class="text"
          >
            Face on the picture matches face on the document
          </p>
          <div>
            <button
              class="leftButton"
            >
              Yes
            </button>
            <button
              class="rightButton"
            >
              No
            </button>
          </div>
        </div>
      </form>
      <button
        class="button"
        disabled=""
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;

exports[`App renders Error view when sumbit fails 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="container"
    >
      <h2>
        Whoops! Something went wrong...
      </h2>
      <div>
        Please 
        <button>
          try again
        </button>
         or contact our 
        <a
          href="https://support.veriff.com/en"
        >
          support.
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`App renders Error view when unable to fetch checks 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="container"
    >
      <h2>
        Whoops! Something went wrong...
      </h2>
      <div>
        Please 
        <button>
          try again
        </button>
         or contact our 
        <a
          href="https://support.veriff.com/en"
        >
          support.
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`App sends submit data when user clicks Submit button and renders success view 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="container"
    >
      <h2>
        Great success!
      </h2>
      <p>
        Your data was submitted successfully.
      </p>
    </div>
  </div>
</div>
`;
